File: PF_Custome_CollectionView/PF_Custome_CollectionView/PFCollectionView.m
//
//  PFCollectionView.m
//  PF_Custome_CollectionView
//
//  Created by 天下林子 on 2019/6/21.
//  Copyright © 2019 linhaoCui. All rights reserved.
//

#import "PFCollectionView.h"

#import "PFCollectionViewFlowLayout.h"
#import "PFCollectionViewCell.h"



#define kScreenW [UIScreen mainScreen].bounds.size.width


@interface PFCollectionView ()<UICollectionViewDelegate, UICollectionViewDataSource, PFCollectionViewFlowLayoutDelegate>

@property (nonatomic, strong) PFCollectionViewFlowLayout  *flowLayout;

@property (nonatomic, strong) NSMutableArray *dataArray;

@property (nonatomic, strong) UICollectionViewFlowLayout  *testFlow;

@end

@implementation PFCollectionView


#pragma mark - Lazy Load

- (PFCollectionViewFlowLayout *)flowLayout
{
    if (!_flowLayout) {

        _flowLayout = [[PFCollectionViewFlowLayout alloc] initWithRowHeight:50];
        _flowLayout.delegate = self;
    }
    
    return _flowLayout;
}


- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        
        self.dataArray = @[@"醉桃花",@"天青色等烟雨,而我在等你",@"而我在等你",@"窈窕淑女",@"君子好逑",@"第一式斩星辰"].mutableCopy;
        //    self.dataArray = @[@"这是一个",@"小普惠化",@"我要还款",@"宽带费",@"这是一个",@"小普惠化",@"我要还款",@"黑色手机宽带费"].mutableCopy;
        //  self.dataArray = @[@"小浦书友会",@"卡面随心随心换换"].mutableCopy;
        
        [self _initUI];
        
    }
    return self;
}


#pragma mark - initUI
- (void)_initUI
{
    UICollectionView *collectionView = [[UICollectionView alloc] initWithFrame:CGRectMake(0, 50, [UIScreen mainScreen].bounds.size.width, 500) collectionViewLayout:self.flowLayout];
    
    collectionView.delegate = self;
    collectionView.dataSource = self;
    
    [self addSubview:collectionView];
    
    [collectionView registerClass:[PFCollectionViewCell class] forCellWithReuseIdentifier:@"cell"];
    
    collectionView.backgroundColor = [UIColor colorWithRed:238/255.0 green:238/255.0 blue:238/255.0 alpha:1.0];
}


#pragma mark - UICollectionViewDataSource

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section
{
	
    return self.dataArray.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    
    PFCollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"cell" forIndexPath:indexPath];
    
    cell.titleLb.text = self.dataArray[indexPath.row];
    
    
    
    if (indexPath.row % 2 == 0)
    {
        cell.backgroundColor = [UIColor orangeColor];
    }
    else
    {
        cell.backgroundColor = [UIColor purpleColor];
    }
    
    return cell;
}


//返回列间距
//MARK:--------返回列间距---------------
-(CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout minimumInteritemSpacingForSectionAtIndex:(NSInteger)section
{
    return 10;
}
//返回行间距
-(CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout minimumLineSpacingForSectionAtIndex:(NSInteger)section
{
    return 25;
}
//返回距离边的距离
-(UIEdgeInsets)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout insetForSectionAtIndex:(NSInteger)section
{
    return UIEdgeInsetsMake(0, 20, 0, 20);
}

//MARK:-设置宽和高度
- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout*)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath
{
    NSString *text = [NSString stringWithFormat:@"   %@   ",self.dataArray[indexPath.row]];
    
    CGSize textMaxSize = CGSizeMake(kScreenW - 2 * 30, MAXFLOAT);
    
    CGFloat ww =  [text boundingRectWithSize:textMaxSize options:NSStringDrawingUsesLineFragmentOrigin attributes:@{NSFontAttributeName : [UIFont systemFontOfSize:15]} context:nil].size.width;
    
    return CGSizeMake(ww, 50);
}

- (BOOL)collectionView:(UICollectionView *)collectionView shouldSelectItemAtIndexPath:(NSIndexPath *)indexPath
{
    return YES;
}

#pragma mark - FlowLayout Delegate

- (CGFloat)obtainItemWidth:(PFCollectionViewFlowLayout *)layout widthAtIndexPath:(NSIndexPath *)indexPath
{
    NSString *text = [NSString stringWithFormat:@"   %@  ",self.dataArray[indexPath.row]];
    CGSize textMaxSize = CGSizeMake(kScreenW - 2 *30, MAXFLOAT);
    CGFloat ww = [text boundingRectWithSize:textMaxSize options:NSStringDrawingUsesLineFragmentOrigin attributes:@{NSFontAttributeName : [UIFont systemFontOfSize:15]} context:nil].size.width;
    
    return ww;
}


@end

File: PF_Custome_CollectionViewCell_swift/PF_Custome_CollectionViewCell/PFCollectionView.swift
//
//  PFCollectionView.swift
//  PF_Custome_CollectionViewCell
//
//  Created by 天下林子 on 2019/6/24.
//  Copyright © 2019 linhaoCui. All rights reserved.
//

import UIKit

class PFCollectionView: UIView {
	
    let dataArray = ["天青色等烟雨","电子书城","全民朗读","大咖主播","活动","直播微课","听单"]
    private lazy var collectionView : UICollectionView = {
        let layout = UICollectionViewFlowLayout.init()
        layout.sectionInset = UIEdgeInsets(top: 10, left: 15, bottom: 0, right: 15)
        layout.minimumLineSpacing = 15
        layout.minimumInteritemSpacing = 15
        layout.itemSize = CGSize(width: 100, height: 40)
        
        let coll = UICollectionView.init(frame: CGRect(x: 0, y: 0, width: UIScreen.main.bounds.size.width, height: UIScreen.main.bounds.size.height - 100), collectionViewLayout: layout)
        coll.delegate = self
        coll.dataSource = self
        coll.backgroundColor = UIColor.white
        coll.register(PFCollectionViewCell.self, forCellWithReuseIdentifier: "cell")
        
        return coll
    }()
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.addSubview(self.collectionView)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
}


extension PFCollectionView : UICollectionViewDelegate, UICollectionViewDataSource {

	func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {

		//return 12
        
        return self.dataArray.count
	}
	
	func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
		
		let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "cell", for: indexPath) as! PFCollectionViewCell
		
		//cell.label.text = "rrrrr"
        
        cell.label.text = self.dataArray[indexPath.row]
        
        if indexPath.row % 2 == 0 {
            
            cell.backgroundColor = UIColor.orange
        }
        else
        {
            cell.backgroundColor = UIColor.red
        }
		return cell
	}
	
	// 每个分区的内边距
	func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, insetForSectionAt section: Int) -> UIEdgeInsets {
		
		return UIEdgeInsets(top: 0, left: 15, bottom: 0, right: 15)
	}
	
	// 最小 item 间距
	func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumInteritemSpacingForSectionAt section: Int) -> CGFloat {
		
		return 15
	}
	
	// 最小行间距
	func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumLineSpacingForSectionAt section: Int) -> CGFloat {
		
		return 15
	}
	
	func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
		
		//return CGSize.init(width: (LBFMScreenWidth - 30) / 3, height:(LBFMScreenWidth - 30) / 3)		
		return CGSize(width: 50, height: 50)
	}
}

extension PFCollectionView : PFCollectionViewFlowLayoutDelegate {
	
    func obtainItemWidth(layout:PFCollectionViewFlowLayout, atIndexPath: IndexPath) -> CGFloat {
        
        //获取cell上字体的宽度
        
        //var widthArr = [Any]()
        var widthArr:[CGFloat] = []
        for item in self.dataArray {
            
            //var text : NSString  = " "
            //text = text.appending(item) as NSString
            let text : NSString  = NSString(string: item)
            
            let maxSize = CGSize(width: UIScreen.main.bounds.size.width - 2 * 30, height: CGFloat(MAXFLOAT))
            
            let ww = NSString(string: text).boundingRect(with: maxSize, options: .usesFontLeading, attributes: [NSAttributedString.Key.font:UIFont.boldSystemFont(ofSize: 20)], context: nil).size.width
            
            widthArr.append(ww)
            
        }
        
        return widthArr[atIndexPath.row]
    }
}
